jps-plugin/src/org/jetbrains/jps/incremental/scala/local/zinc/IntellijExternalLookup.scala

```diff
@@ -0,0 +1,51 @@
+package org.jetbrains.jps.incremental.scala.local.zinc
+
+import java.io.File
+import java.util
+import java.util.Optional
+
+import org.jetbrains.jps.incremental.scala.Client
+import org.jetbrains.jps.incremental.scala.data.CompilationData
+import sbt.internal.inc._
+import xsbti.compile._
+
+import scala.collection.JavaConverters._
+
+case class IntellijExternalLookup(compilationData: CompilationData, client: Client, isCached: Boolean)
+  extends ExternalLookup {
+
+  private val all = compilationData.zincData.allSources.toSet
+  private val changedSources = compilationData.sources.toSet
+
+  override def hashClasspath(classpath: Array[File]): Optional[Array[FileHash]] =
+    Optional.of(classpath.map(f => FileHash.of(f, f.hashCode())))
+
+  override def changedSources(previousAnalysis: CompileAnalysis): Option[Changes[File]] =
+    if (isCached) None else {
+    val previousSources = previousAnalysis.readStamps().getAllSourceStamps.keySet().asScala.toSet
+
+    Some(new UnderlyingChanges[File] {
+      def added: Set[File] = all -- previousSources
+
+      def removed: Set[File] = previousSources -- all
+
+      def changed: Set[File] = changedSources & previousSources
+
+      def unmodified: Set[File] = previousSources -- changedSources
+    })
+  }
+
+  override def changedBinaries(previousAnalysis: CompileAnalysis): Option[Set[File]] = Some(Set())
```

Comment:
This is our optimization to speed up small compilation (this may take a lot of time). It can by turn off (if we return here empty) but in such case please allow us to override this (e.g. in our plugin).

Same thing applies to `override def removedProducts(previousAnalysis: CompileAnalysis): Option[Set[File]] = Some(Set())`.